crates/nargo_cli/src/cli/compile_cmd.rs

```diff
@@ -120,18 +131,18 @@ pub(crate) fn run<B: Backend>(
 
 pub(crate) fn compile_circuit<B: Backend>(
     backend: &B,
-    package: Option<String>,
-    program_dir: &Path,
+    context: &mut Context,
+    crate_id: CrateId,
     compile_options: &CompileOptions,
-) -> Result<(CompiledProgram, Context), CliError<B>> {
-    let (mut context, crate_id) = resolve_root_manifest(program_dir, package)?;
-    let result = compile_main(&mut context, crate_id, compile_options);
-    let mut program = report_errors(result, &context, compile_options.deny_warnings)?;
-
+) -> Result<CompiledProgram, ReportedErrors> {
+    let result = compile_main(context, crate_id, compile_options);
+    let mut program = report_errors(result, context, compile_options.deny_warnings)?;
     // Apply backend specific optimizations.
     let (optimized_circuit, opcode_labels) = optimize_circuit(backend, program.circuit)
         .expect("Backend does not support an opcode that is in the IR");
 
+    // TODO: Why does this set `program.circuit` to `optimized_circuit` instead of the function taking ownership
```

Comment:
@TomAFrench What was the purpose of this design?

Reply:
erm, no amazingly smart reasons tbh. Feel free to rework this as we're going to continue separating compilation and optimization as time goes on.

Reply:
I'll open a follow-up issue

Reply:
#2110